torch/distributed/pipelining/schedules.py

```diff
@@ -656,12 +663,19 @@ def _step_microbatches(
         stage_index_to_stage: Dict[int, _PipelineStageBase] = {
             stage.stage_index: stage for stage in self._stages
         }
-        prev_rank: int = (self.rank - 1) % self.pp_group_size
-        next_rank: int = (self.rank + 1) % self.pp_group_size
+
+        # determine prev_rank and next_rank based on which ranks are next to
+        # the stages in the pipeline_order
+        all_prev_ranks: Set[int] = set()
+        all_next_ranks: Set[int] = set()
+        for stage_index in stage_index_to_stage.keys():
+            # TODO: assumption that stages only communicate from distances of +1/-1 (no skip connections)
```

Comment:
you're being a bit uncommitted i think- should we generalize the code to support skip connections?  

if so we need tests for that, and we need a solution to this issue - how to determine who the other ranks are at this point in the code?

or if not, maybe its not great to partialy generalize the code below (making for-loops e.g. `for prev_rank in all_prev_ranks:`

perhaps better to fully punt on 'skip connections' till we can fully test it and ensure its free of edge cases etc.  and only land the smaller set of changes needed to enable decoupling rank from stage id.

Reply:
I wanted to punt on skip connections as well, but my hope is that if the changes in this PR are landed then it should be easy to add as we really need only need to update the stage API for users to pass in output arg to stage mappings, and then update this area of the code you commented on.

The for-loop is necessary regardless of skip connections due to the reason mentioned in the previous comment.

Reply:
wait.. is the for loop even being tested? my read of this was that all_prev_ranks starts out empty, and we just add one rank to it which is the rnak for stage_index-1.

where do we add a second rank to all_prev_ranks?

if its just one rank, then the for loop is just running the same way as it used to when it was not a for loop isn't it?

Reply:
In the test case the rank to stages are: 

rank0: stage0, stage3
rank1: stage1, stage 2

Since we just check stage_index-1, for rank 1 all_prev_rank would be [0, 1] (it includes itself), so the loop does execute.

Reply:
i think i was confused that we were setting 'all_prev_ranks' for multiple stages on this rank at the same time.  i missed your change above for adding a for loop over stage id.

i think this might actually be a risk.  let me add a comment there